File: ports/go/src/network/socket/bcstcpsocket.go
package socket

import (
	"bufio"
	"net"
)

// BcsTcpSocket abstracts a network client connection
type BcsTcpSocket struct {
	conn net.Conn
}

func NewBcsTcpSocket() *BcsTcpSocket {
	return &BcsTcpSocket{}
}

func (s *BcsTcpSocket) ConnectToHost(host string, port string) error {
	conn, err := net.Dial("tcp", host+":"+port)
	if err != nil {
		return err
	}
	s.conn = conn
	return nil
}

func (s *BcsTcpSocket) Write(data []byte) (int, error) {
	return s.conn.Write(data)
}

func (s *BcsTcpSocket) ReadLine() (string, error) {
	reader := bufio.NewReader(s.conn)
	return reader.ReadString('\n')
}

func (s *BcsTcpSocket) Disconnect() error {
	if s.conn != nil {
		return s.conn.Close()
	}
	return nil
}

File: ports/rust/src/network/socket/bcstcpsocket.rs
use std::io::{self, BufRead, BufReader, Write};
use std::net::TcpStream;

// BcsTcpSocket abstracts a network client connection for Rust
pub struct BcsTcpSocket {
    stream: Option<TcpStream>,
}

impl BcsTcpSocket {
    pub fn new() -> Self {
        BcsTcpSocket { stream: None }
    }

    pub fn connect_to_host(&mut self, host: &str, port: &str) -> io::Result<()> {
        let address = format!("{}:{}", host, port);
        let stream = TcpStream::connect(address)?;
        self.stream = Some(stream);
        Ok(())
    }

    pub fn write(&mut self, data: &[u8]) -> io::Result<usize> {
        if let Some(ref mut stream) = self.stream {
            stream.write(data)
        } else {
            Err(io::Error::new(io::ErrorKind::NotConnected, "Socket not connected"))
        }
    }

    pub fn read_line(&mut self) -> io::Result<String> {
        if let Some(ref mut stream) = self.stream {
            let mut reader = BufReader::new(stream);
            let mut buffer = String::new();
            reader.read_line(&mut buffer)?;
            Ok(buffer)
        } else {
            Err(io::Error::new(io::ErrorKind::NotConnected, "Socket not connected"))
        }
    }

    pub fn disconnect(&mut self) -> io::Result<()> {
        self.stream = None; // Drops the connection
        Ok(())
    }
}
